src/main/java/ca/fxco/configurablepistons/blocks/pistons/basePiston/BasicPistonBaseBlock.java

```diff
@@ -299,4 +294,218 @@ public void playEvents(Level level, GameEvent event, BlockPos pos) {
         );
         //level.gameEvent(event, pos);
     }
+
+    /**
+     * Returns whether this piston can move the given block,
+     * taking into account world height/border as well as
+     * vanilla and custom piston push reactions.
+     */
+    public boolean canMoveBlock(BlockState state, Level level, BlockPos pos, Direction moveDir, boolean allowDestroy, Direction pistonFacing) {
+        // coordinate related checks (world height/world border)
+
+        if (level.isOutsideBuildHeight(pos) || !level.getWorldBorder().isWithinBounds(pos))
+            return false;
+        if (state.isAir())
+            return true; // air is never in the way
+        if (moveDir == Direction.DOWN && pos.getY() == level.getMinBuildHeight())
+            return false;
+        if (moveDir == Direction.UP && pos.getY() == level.getMaxBuildHeight() - 1)
+            return false;
+
+
+        // piston push reaction/ custom piston behavior
+
+        ConfigurablePistonBehavior customBehavior = (ConfigurablePistonBehavior)state.getBlock();
+
+        if (customBehavior.usesConfigurablePistonBehavior()) { // This is where stuff gets fun
+            if (!customBehavior.isMovable(state))
+                return false;
+            if (moveDir == pistonFacing) {
+                if (!customBehavior.canPistonPush(state, moveDir))
+                    return false;
+            } else {
+                if (!customBehavior.canPistonPull(state, moveDir))
+                    return false;
+            }
+            if (!allowDestroy && !customBehavior.canDestroy(state))
+                return false;
+        } else {
+            if (state.is(ModTags.UNPUSHABLE) || state.getDestroySpeed(level, pos) == -1.0F)
+                return false;
+            if (state.is(ModTags.PISTONS) && state.getValue(EXTENDED))
+                return false;
+            switch (state.getPistonPushReaction()) {
+                case BLOCK -> { return false; }
+                case DESTROY -> {
+                    if (!allowDestroy)
+                        return false;
+                }
+                case PUSH_ONLY -> {
+                    if (moveDir != pistonFacing)
+                        return false;
+                }
+                default -> { }
+            }
```

Comment:
[redacted]

I see that you are making it so that `canMoveBlock` also runs for all checks not just `default`

Reply:
Yes. The idea was that the main `canMoveBlock` method does the world height/border checks, as well as making sure the given block has the right push reaction, while the second `canMoveBlock` method allows for pistons to have custom behavior, e.g. vanilla pistons cannot move block entities, but you could make a special piston that ignores another type of block, like a piston that cannot move other pistons or whatever.

Reply:
Why not use the API right above it instead?